raiden/utils/__init__.py

```diff
@@ -288,3 +289,36 @@ def is_frozen():
 
 def event_decoder(event: Log, contract_translator: ContractTranslator):
     return contract_translator.decode_event(event.topics, event.data)
+
+
+ResultCache = namedtuple('ResultCache', ('result', 'timestamp'))
+
+
+def cache_response_timewise(seconds=600):
```

Comment:
I really dislike the assumption about the attributes of the wrapped object, `wrappingobj.lock`.

I suggest declaring/instantiating the lock, timestamp and result vairables inside the `_cache_response_timewise`, and using the `nonlocal` keyword. The function will close over the objects, the same way as in javascript, so the `results_cache` and `lock` don't need to be instances of the wrapped object, these values can be instantiated inside the decorator itself.

Example:

```
>>> def a():
...   b = 0
...   def c():
...     nonlocal b
...     b += 1
...     return b
...   return c
...
>>> c = a()
>>> c()
1
>>> c()
2
>>> c()
3
```

Additionally, please use `functools.wraps` to mantain the function metadata

Reply:
Question: Would it not be better to make a class for this .. say ... `CacheResponse` which will act as the interface for objects which want to use `cache_response_timewise()`?

This way `rpc.client` and any other object which wants to implement this interface can inherit from this new class.

If I understand correctly how the `nonlocal` keyword functions, all uses of this cache would use the same lock and this may not be appropriate. If both the rpc.client and another object want to have cached responses then we want them to use a different lock each.

Reply:
> CacheResponse which will act as the interface for objects which want to use cache_response_timewise()?

Well, my problem is that the decorator is not self-contained. With that approach using the decorator forces the wrapped object to change, I don't think the interface is any different to some predefine attributes in that regard.

> all uses of this cache would use the same lock and this may not be appropriate.

That not how it works. The same closure will use the same objects, but different closures will have different objects:
```python
>>> def a():
...   b = 0
...   def c():
...     nonlocal b
...     b += 1
...     return b
...   return c
...
>>> c = a()
>>> c()
1
>>> c()
2
>>> c()
3
>>> c2 = a()
>>> c2()
1
>>> c2()
2
>>> c2()
3
```

Reply:
@hackaugusto So the resulting lock would no longer be per object, but per attribute accessed? So a different lock for `gasprice` and a different lock for `gaslimit`, is that correct?

In our case and since it's only 2 attributes this may be indeed fine. But how could you do the same for when you need to have 1 lock per object but not per attribute?

Reply:
> So the resulting lock would no longer be per object, but per attribute accessed? So a different lock for gasprice and a different lock for gaslimit, is that correct?

Yes, if the lock is instantiated inside the decorator.

> In our case and since it's only 2 attributes this may be indeed fine. But how could you do the same for when you need to have 1 lock per object but not per attribute?

I don't think the lock should go inside the decorato. There are two different concerns here, the synchronization and the caching.

I suggested only having the caching as a concern for the decorator, and then you can do something else for the locking, like having a lock that is an attribute of the class and using the same lock for different methods.